integrations: Add SonarQube webhook integration.

tushar912 opened the pull request:
Fixes #13395

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
tested manually as well as using ./tools/test-backend zerver/webhooks/sonarqube

timabbott commented on zerver/webhooks/sonarqube/fixtures/success.json:
```diff
@@ -0,0 +1,64 @@
+    "url": "http://localhost:9000/dashboard?id=test-sonar"
+  },
+  "branch": {
+    "name": "master",
```
Can you work the branch name into the message?  I think it's likely to be important for any project that has multiple branches.

tushar912 replied:
Ok.

timabbott commented on zerver/webhooks/sonarqube/tests.py:
```diff
@@ -0,0 +1,53 @@
+        expected_topic = "test-sonar"
+
+        expected_message = """
+Project [test-sonar](http://localhost:9000/dashboard?id=test-sonar) analysis resulted in error:
```
Maybe just "... analysis of branch {branch} resulted" as how to work the branch name in?

tushar912 replied:
Sure

timabbott commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,89 @@
+    )
+    msg += conditions
+
+    return msg
```
I think we probably want to call this function `render_body`?

tushar912 replied:
Sure

tushar912 commented:
@timabbott I have made the changes.

eeshangarg commented on zerver/webhooks/sonarqube/doc.md:
```diff
@@ -0,0 +1,15 @@
+1. {!create-stream.md!}
+
+1. {!create-bot-construct-url-indented.md!}
+   By default, the integration will use the project's name as the topic.
```
I honestly do not think we need this line about the topic, so would recommend just removing it.

tushar912 replied:
Sure

eeshangarg commented on zerver/webhooks/sonarqube/doc.md:
```diff
@@ -0,0 +1,15 @@
+   By default, the integration will use the project's name as the topic.
+
+1. On SonarQube Select **Administration** either in your project or on global level.
+   Select **Webhooks** (on global level **Configurations > Webhooks**). Click **Create**.
```
I would rephrase this one slightly, like so:
1. To configure webhooks for a specific SonarQube project, go to the project and select **Administration**. Select **Webhooks** and click **Create**. **Note**: you can also configure webhooks globally by going to **Configurations -> Webhooks** in SonarQube.

tushar912 replied:
Ok

eeshangarg commented on zerver/webhooks/sonarqube/doc.md:
```diff
@@ -0,0 +1,15 @@
+1. On SonarQube Select **Administration** either in your project or on global level.
+   Select **Webhooks** (on global level **Configurations > Webhooks**). Click **Create**.
+
+1. Choose **Name** for webhook, set **URL** to the URL constructed above and click **Create**.
```
Change this to:
1. Set **Name** to a name for the webhook. Set **URL** to the URL constructed above and click **Create**.

tushar912 replied:
Ok

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+
+inverse_operators = {
+    "WORSE_THAN": "should be better or equal to",
+    "GREATER_THAN": "should be less or equal to",
```
Change to `should be less than or equal to`

tushar912 replied:
Ok

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+inverse_operators = {
+    "WORSE_THAN": "should be better or equal to",
+    "GREATER_THAN": "should be less or equal to",
+    "LESS_THAN": "should be greater or equal to",
```
Change to `should be greater than or equal to`.

tushar912 replied:
Ok

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+    conditions = payload["qualityGate"]["conditions"]
+    conditions = parse_conditions(conditions)
+    if branch:
+        if conditions == "":
```
This can just be `if not conditions:`

tushar912 replied:
Ok

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+
+        return msg
+
+    if conditions == "":
```
Same as above.

tushar912 replied:
Ok

eeshangarg reviewed (changes requested):
@tushar912 Hey! Thanks for working on this! This PR looks pretty good, I just left a few minor comments and suggestions that I would like you to address. Please feel to reach out if you have any questions. Thanks! :)

tushar912 commented:
@timabbott @eeshangarg I have made the changes.

timabbott commented:
I'll wait for @eeshangarg to comment before reviewing again.

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+from zerver.lib.webhooks.common import check_send_webhook_message
+from zerver.models import UserProfile
+
+inverse_operators = {
```
Since this is technically a constant, we should probably capitalize `inverse_operators`.

tushar912 replied:
Ok

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+    "LESS_THAN": "should be greater than or equal to",
+}
+
+templates = {
```
Since this is technically a constant, we should capitalize `templates`.

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+                project_name, project_url, branch, quality_gate_status
+            )
+        msg = "Project [{}]({}) analysis of branch {} resulted in {}:\n".format(
+            project_name, project_url, branch, quality_gate_status
```
Since we are using these message formatting templates multiple times, I would recommend moving these to their constant at the top of the file as well. E.g. names "MESSAGE_WITH_CONDITIONS", "MESSAGE_WITHOUT_CONDITIONS".

tushar912 replied:
Sure

eeshangarg commented on zerver/webhooks/sonarqube/view.py:
```diff
@@ -0,0 +1,104 @@
+    user_profile: UserProfile,
+    payload: Dict[str, Any] = REQ(argument_type="body"),
+) -> HttpResponse:
+    topic = payload["project"]["name"]
```
Another potential improvement here is to include the branch name in the topic. That would lead to messages for the same branch being grouped under the same topic and simplifying a lot of the logic in `render_body` as well. Ideally, you want to easily be able to see the analysis notifications for `master` under the same topic. This can be a part of a follow-up PR as well!

eeshangarg replied:
@timabbott Not sure if you have any thoughts on this. In general, grouping messages by branch seems to be something we do a lot in our integrations. I think it is a good principle to adhere to consistently across integrations!

tushar912 replied:
I added the branch to topic similar to what we do for github integration. Here is the screenshot
![001](https://user-images.githubusercontent.com/59223034/115963553-6de14500-a53d-11eb-8548-e581feaab7aa.png)

eeshangarg reviewed (changes requested):
Hey @tushar912! Thanks for addressing my feedback, I left a few more suggestions for improvement that I think would clear up the structure of the code a little bit. Please feel free to reach out if you have any further questions! Thanks! :)

tushar912 commented:
@eeshangarg Sorry for the delay on this. I have made the following changes based on your suggestions
* Capitalized constants.
* Added templates for different messages.
* Added branch to topic if present.
* Split the render_body function into 2 according to branch present.

eeshangarg reviewed (approved):
Thanks for working on this, @tushar912! This looks good to me. :) @timabbott I think we should merge this and then perhaps make follow-up improvements if need be. Thanks!

jwillmer commented:
@eeshangarg would you mind to merge it. I have a use for this integration 😉

timabbott commented:
Sorry we lost track of this, and thanks for the bump @jwillmer!  Merged, thanks @tushar912 for writing this!

timabbott commented:
@jwillmer since this is a new integration, please be especially encouraged to open issues or start a thread in #integrations if you have any feedback on how we might improve the formatting. 

(If you're self-hosting, it should be easy to cherry-pick; it should be on Zulip Cloud in a few days).
